lightbeam/validate.py

```diff
@@ -65,6 +66,13 @@ def validate(self):
                 # to comparatively small datasets (sections, schools, students).
                 self.build_local_reference_cache(endpoint)
             asyncio.run(self.validate_endpoint(endpoint))
+        
+        # write structured output (if needed)
+        self.lightbeam.write_structured_output("validate")
```

Comment:
The only problem I see is that this won't execute if the fast-fail threshold is hit, causing early exit. I think we might need to rethink how we exit during a fast fail. It's tricky to handle this given the multiple async tasks running together but I'm pondering what we could do…

Reply:
My vote here would be to have some sort of shutdown method* that is the single source of truth for exiting. Instead of calls to `logger.critical` any premature exits can instead call the class's shutdown method, which can guarantee that any cleanup tasks, including possible logging to disk, are executed. What do you think?

*ideally at the lightbeam level, but may need to implemented method-by-method for now?

Reply:
I've made a few updates that should resolve this. I tested it and the results-file is output even when we fail-fast, so marking this resolved. Thanks for raising the issue!